boa/src/syntax/parser/mod.rs

```diff
@@ -105,27 +119,25 @@ pub type ParseResult = Result<Node, ParseError>;
 
 #[derive(Debug)]
 pub struct Parser {
-    /// The tokens being input
-    tokens: Vec<Token>,
-    /// The current position within the tokens
-    pos: usize,
+    /// Cursor in the parser, the internal structure used to read tokens.
+    cursor: Cursor,
 }
 
 macro_rules! expression { ( $name:ident, $lower:ident, [ $( $op:path ),* ] ) => {
     fn $name (&mut self) -> ParseResult {
         let mut lhs = self. $lower ()?;
-        while let Ok(tok) = self.peek_skip_lineterminator() {
+        while let Some(tok) = self.peek_skip_lineterminator().cloned() {
```

Comment:
Could we somehow avoid this? It seems that Rust gives a "multiple mutable reference" error when you return a read-only reference from a mutable reference more than once. Is this something the compiler should understand it's safe?

Reply:
Hmm without checking this out and playing with it I’m not sure. Will need to take a look.